### PR TITLE
DAOS-12706 object: Some memory corruption issue

### PR DESCRIPTION
**Pull request by wangdi1**
Fix some memory corruption issuse for csum in
obj_get_iods_offs_by_oid().

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by wangdi1** [comment]
failure due to DAOS-12857

**Comment by wangdi1** [comment]
@Nasf-Fan @liuxuezhao  please review.

File: src/object/srv_obj.c [hunk 1]
```diff
@@ -1539,8 +1539,7 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc,
 		if (ioc->ioc_coc->sc_props.dcp_csum_enabled) {
 			rc = csum_add2iods(ioh,
 					   orw->orw_iod_array.oia_iods,
-					   orw->orw_iod_array.oia_iod_nr,
-					   ioc->ioc_coc->sc_csummer,
+					   iods_nr, ioc->ioc_coc->sc_csummer,
```
Comment (Nasf-Fan) [inline src/object/srv_obj.c:1542]: why "orw->orw_iod_array.oia_iods" instead of the input "iods"? that is the pairs of "iods_nr".

Comment (Nasf-Fan) [reply src/object/srv_obj.c:1542]: So we may add "D_ASSERT(orw->orw_iod_array.ioa_iods should be always == iods)" ? if has other chance to refresh the patch?

Anyway, your patches seems not resolve the issue:

https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11624/2/artifact/Functional%20Hardware%20Medium/scrubber/target_auto_eviction.py/stacktraces.wolf-138/core.daos_io_1.1678316154.114222.stacktrace/*view*/

Thread 1 (Thread 0x7f176168f700 (LWP 114471)):
#0  0x00007f17dfe5fa9f in raise () from /lib64/libc.so.6
No symbol table info available.
#1  0x00007f17dfe32e05 in abort () from /lib64/libc.so.6
No symbol table info available.
#2  0x00007f17dfe32cd9 in __assert_fail_base.cold.0 () from /lib64/libc.so.6
No symbol table info available.
#3  0x00007f17dfe583f6 in __assert_fail () from /lib64/libc.so.6
No symbol table info available.
#4  0x0000559e25a607cd in proc_struct_dcs_csum_info_adv (proc=proc@entry=0x7f175ea29f10, proc_op=proc_op@entry=CRT_PROC_ENCODE, csum=0x7f175ed16810, idx=idx@entry=0, nr=<optimized out>) at src/object/rpc_csum.c:65
        buf_len = 2
        rc = 0
        __func__ = "proc_struct_dcs_csum_info_adv"
        __PRETTY_FUNCTION__ = "proc_struct_dcs_csum_info_adv"
#5  0x0000559e25a60d4b in proc_struct_dcs_csum_info (proc=proc@entry=0x7f175ea29f10, proc_op=proc_op@entry=CRT_PROC_ENCODE, csum=<optimized out>) at src/object/rpc_csum.c:95
No locals.
#6  0x0000559e25a60de3 in proc_struct_dcs_csum_info (csum=<optimized out>, proc_op=CRT_PROC_ENCODE, proc=0x7f175ea29f10) at src/object/rpc_csum.c:92
No locals.
#7  crt_proc_struct_dcs_csum_info (proc=proc@entry=0x7f175ea29f10, proc_op=<optimized out>, p_csum=p_csum@entry=0x7f175ea8b958) at src/object/rpc_csum.c:110
        csum_enabled = true
        rc = 0
        __func__ = "crt_proc_struct_dcs_csum_info"
#8  0x0000559e25a5190e in crt_proc_obj_rw_in (proc=0x7f175ea29f10, ptr=0x7f175ea8b8b0) at src/object/obj_rpc.c:1079
        proc_op = CRT_PROC_ENCODE
        rc = 0
        __func__ = "crt_proc_obj_rw_in"

**Review by liuxuezhao — COMMENTED** [review]
just test "orterun -np 1 ./daos_test -I -u 17 -n ./daos_control.yml --csum_type "crc32" " cannot work even after this PR, probably still some problem in singv csum handling.

File: src/object/srv_obj.c (hunk 1 above)
Comment (wangdi1) [reply src/object/srv_obj.c:1542]: This csum assertion is different issue DAOS-12857.

File: src/object/srv_obj.c [hunk 2]
```diff
@@ -1736,20 +1728,24 @@ obj_get_iods_offs_by_oid(daos_unit_oid_t uoid, struct obj_iod_array *iod_array,
 			(*iods)[idx].iod_recxs = &iod_parent->iod_recxs[siod->siod_idx];
 			(*iods)[idx].iod_nr = siod->siod_nr;
 			if (csums != NULL) {
-				(*csums)[i].ic_data = &iod_pcsum->ic_data[siod->siod_idx];
-				(*csums)[i].ic_nr = siod->siod_nr;
+				(*csums)[idx].ic_data = &iod_pcsum->ic_data[siod->siod_idx];
+				(*csums)[idx].ic_nr = siod->siod_nr;
 			}
 		} else {
 			if (iod_parent->iod_recxs != NULL)
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/object/srv_obj.c:1735]: ah, hmm,   I checked the original obj_ec_rw_req_split(), which does not include this iod_recxs check under SINGLE branch, so I had thought iod_recxs is being reused for some cases.  ok, I can remove this check if the patch needs to be refreshed.

**Review by Nasf-Fan — APPROVED** [review]

**Comment by wangdi1** [comment]
failure due to daos-12706

**Pull request merged**